cdap-gateway/src/main/java/co/cask/cdap/gateway/router/handlers/OutboundHandler.java

```diff
@@ -42,15 +40,14 @@ public OutboundHandler(Channel inboundChannel) {
   }
 
   @Override
-  public void messageReceived(final ChannelHandlerContext ctx, final MessageEvent e) throws Exception {
-    ChannelBuffer msg = (ChannelBuffer) e.getMessage();
-
-    Channels.write(inboundChannel, msg).addListener(new ChannelFutureListener() {
-      @Override
-      public void operationComplete(ChannelFuture future) throws Exception {
-        OutboundHandler.super.messageReceived(ctx, e);
-      }
-    });
+  public void messageReceived(ChannelHandlerContext ctx, MessageEvent event) throws Exception {
+    ChannelBuffer msg = (ChannelBuffer) event.getMessage();
+    int readerIndex = msg.readerIndex();
+    int writerIndex = msg.writerIndex();
+    super.messageReceived(ctx, event);
+    // write the same channel buffer to inbound channel
+    msg.setIndex(readerIndex, writerIndex);
```

Comment:
The `super.messageReceived` would have read the channel buffer, thus moving the read index to the end. If you read the channel buffer in that state, it will be empty. Hence reset the read pointer to earlier state, so that we can read it again.